Alert configuration resource and data source

PacoDw opened the pull request:
Added:

  - mongodbatlas_alert_configuration resource.
  - mongodbatlas_alert_configuration resource acceptance testing.
  - mongodbatlas_alert_configuration resource website documentation.
  - mongodbatlas_alert_configuration data source.
  - mongodbatlas_alert_configuration data source acceptance testing.
  - mongodbatlas_alert_configuration data source website documentation.

#### Basic Example Configuration
```hcl
resource "mongodbatlas_alert_configuration" "test" {
  project_id = "<PROJECT-ID>"
  event_type = "OUTSIDE_METRIC_THRESHOLD"
  enabled    = true

  notification {
    type_name     = "GROUP"
    interval_min  = 5
    delay_min     = 0
    sms_enabled   = false
    email_enabled = true
  }

  matcher {
    field_name = "HOSTNAME_AND_PORT"
    operator   = "EQUALS"
    value      = "SECONDARY"
  }

  metric_threshold = {
    metric_name = "ASSERT_REGULAR"
    operator    = "LESS_THAN"
    threshold   = 99.0
    units       = "RAW"
    mode        = "AVERAGE"
  }
}


data "mongodbatlas_alert_configuration" "test" {
	project_id             = "${mongodbatlas_alert_configuration.test.project_id}"
	alert_configuration_id = "${mongodbatlas_alert_configuration.test.alert_configuration_id}"
}
```
#### Example Configuration with multiple notifications
```hcl
resource "mongodbatlas_alert_configuration" "test" {
  project_id = "<PROJECT-ID>"
  event_type = "NO_PRIMARY"
  enabled    = true

  notification {
    type_name     = "GROUP"
    interval_min  = 5
    delay_min     = 0
    sms_enabled   = true
    email_enabled = false
  }

  notification {
    type_name     = "ORG"
    interval_min  = 5
    delay_min     = 1
    sms_enabled   = false
    email_enabled = true
  }
}
```
#### Example Configuration with multiple matchers
```hcl
resource "mongodbatlas_alert_configuration" "test" {
  project_id = "%s"
  event_type = "NO_PRIMARY"
  enabled    = true

  notification {
    type_name     = "GROUP"
    interval_min  = 5
    delay_min     = 0
    sms_enabled   = false
    email_enabled = true
  }

  matcher {
    field_name = "TYPE_NAME"
    operator   = "EQUALS"
    value      = "SECONDARY"
  }

  matcher {
    field_name = "TYPE_NAME"
    operator   = "CONTAINS"
    value      = "MONGOS"
  }
} 
```

PacoDw committed:
chore: updated vendor

chore: updated vendor and dependencies

PacoDw committed:
feat: added alert configuration resource

PacoDw committed:
feat: added alert configuration test resource

PacoDw committed:
feat: added alert configuration data source

PacoDw committed:
feat: added alert configuration test data source

PacoDw committed:
feat: added alert configuration resource and datasource inside of the provider

PacoDw committed:
website: added documentation of the alert configuration resource

PacoDw committed:
website: added documentation of the alert configuration data source

PacoDw committed:
website: added resource and datasource documentation

themantissa reviewed (changes requested):
522 files!  Please reduce to files to be reviewed.

PacoDw commented:
@themantissa it's the vendor file and dependencies that were updated, just omit the first two commits:
* chore: updated vendor `0442562`
* chore: updated vendor and dependencies `42fc0b1`

PacoDw committed:
chore: changed client functions due to updated vendor

themantissa commented on mongodbatlas/data_source_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,229 @@
+				ForceNew: true,
+				Required: true,
+			},
+			"group_id": {
```
project id is the same as group id - just checking, is this to accept the returned group_id -  i.e. https://docs.atlas.mongodb.com/reference/api/alert-configurations-create-config/ the group id returned is the same used in the call?

PacoDw replied:
Yes, it is the same I can remove it if you prefer that, I decided to leave it due to the documentation.

themantissa replied:
Ah yes, it's in the example in the documentation isn't it?  I would remove it as we already know the value and having the same twice might prove confusing.

PacoDw replied:
okay make sense, thanks I will remove it 👍

marinsalinas replied:
@PacoDw, will you remove it?

themantissa commented on mongodbatlas/resource_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,523 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"group_id": {
```
Same double check as for the data source above, this is because we both use the project id and return it with this method?

PacoDw replied:
The same to the above comment

themantissa replied:
Same response as above, using both terms will likely confuse users.  Will see if I can get our docs updated.

themantissa commented on mongodbatlas/resource_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,523 @@
+	var err error
+
+	// TO update is nesessary to send the original create alert configuration request if not the server returns
+	// error 500
```
Just a bit of clarification (ensure I've kept the meaning though):
// In order to update an alert config it is necessary to send the original alert configuration request again, if not the //server returns an error 500

PacoDw replied:
Thank you so much ! 👍

themantissa commented on mongodbatlas/resource_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,523 @@
+	if err != nil {
+		return fmt.Errorf(errorReadAlertConf, err)
+	}
+	// Removing the computed attributest to create the original create request
```
Adjustment:
// Removing the computed attributes to recreate the original request

themantissa commented on mongodbatlas/resource_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,523 @@
+	// If something change, always the matches need to be sent like
+	// the Terraform state to due that the req variable got above doesn't have
+	// the "field_name" in each matcher item attribute so if the request is sent like
+	// above we will get an error
```
// If matchers changes ensure we are sending the information like the Terraform state
// because the req variable above doesn't have the "field_name" in each matcher item attribute
// if sent as is, the server sends an error

PacoDw replied:
👍

themantissa commented on mongodbatlas/resource_mongodbatlas_alert_configuration.go:
```diff
@@ -0,0 +1,523 @@
+	}
+
+	// it's necessary if just enabled attr is seated, due if only the enabled attr is sent the server returns error 500,
+	// so we need to use the enabled/disabled entry point to set it
```
I think what you are trying to describe is the following - let me know if it makes sense:
// Cannot enable/disable ONLY via update (if only send enable as changed field server returns a 500 error) so have to use different method to change enabled.

PacoDw replied:
it's the same but better, thanks 👍

themantissa commented on website/docs/d/alert_configuration.html.markdown:
```diff
@@ -0,0 +1,217 @@
+
+# mongodbatlas_alert_configuration
+
+`mongodbatlas_alert_configuration` describes a Alert Configuration.
```
just change a to an, that is 
"describes an Alert Configuration"

themantissa commented on website/docs/r/alert_configuration.html.markdown:
```diff
@@ -0,0 +1,219 @@
+
+# mongodbatlas_alert_configuration
+
+`mongodbatlas_alert_configuration` provides an Alert Configuration resource for the project associated to {GROUP-ID}
```
slight modification:
provides an Alert Configuration resource to define the conditions that trigger an alert and the methods of notification within a MongoDB Atlas project.

themantissa commented on website/docs/r/alert_configuration.html.markdown:
```diff
@@ -0,0 +1,219 @@
+## Argument Reference
+
+* `project_id` - (Required) The ID of the project where the alert configuration will create.
+* `enabled` - If omitted, the configuration is disabled.
```
True for Terraform?  Doesn't it have to be true or false for the resource?

PacoDw replied:
yes and it is, but If the attribute is ommited on terraform, the configuration will disabled on the API. I will extend the comment to be more descriptive.

themantissa commented on website/docs/d/alert_configuration.html.markdown:
```diff
@@ -0,0 +1,217 @@
+* `group_id` - Unique identifier of the project that owns this alert configuration.
+* `created` - Timestamp in ISO 8601 date and time format in UTC when this alert configuration was created.
+* `updated` - Timestamp in ISO 8601 date and time format in UTC when this alert configuration was last updated.
+* `enabled` - If omitted, the configuration is disabled.
```
For the data source I think this needs to be different?   Per the Atlas documentation for a get "If set to true, the alert configuration is enabled. If enabled is not specified in a POST command, it defaults to false."  So will the Terraform datasource will only export enabled if it's true?  I would think the way this works (if I'm understanding it right) is Terraform would see it as true/false but the underlying API would just omit it on a read?

PacoDw replied:
okay let me change that, thanks @themantissa

themantissa reviewed (changes requested):
A few comments/questions and grammar edits but nothing huge, thank you!

themantissa commented:
Regarding the 522 files - I did limit it to the commits you noted but it still made it challenging to review in that jira kept wanting to go back to showing all 522 w/ any little change - at one point I couldn't find my already made comments.   So please, for my sanity, try to ensure we don't have another PR with vendor included.  Thank you!

PacoDw committed:
chore: improved code comments

PacoDw committed:
chore: updated docuemtantions

themantissa commented on website/docs/d/alert_configuration.html.markdown:
```diff
@@ -0,0 +1,217 @@
+* `group_id` - Unique identifier of the project that owns this alert configuration.
+* `created` - Timestamp in ISO 8601 date and time format in UTC when this alert configuration was created.
+* `updated` - Timestamp in ISO 8601 date and time format in UTC when this alert configuration was last updated.
+* `enabled` - If set to true, the alert configuration is enabled.If enabled is not specified in a POST command, it defaults to false.
```
Not sure POST makes sense here - data sources info?      Maybe change to:
If set to true, the alert configuration is enabled.  If enabled is not exported it is set to false.

PacoDw replied:
Thank you!

themantissa reviewed (approved):
3 small things - then can approve.

themantissa reviewed (changes requested):
See my comments in the approve, should actually have been request changes.  (Sorry super tired)

PacoDw committed:
chore: group_id was removed and fixing issues due to the API changes

PacoDw committed:
website: improved data source documentation

themantissa reviewed (approved):
LGTM, except for conflicts. Please resolve first.

marinsalinas reviewed (approved):
LGTM  just want to check if you will remove the group_id argument!

marinsalinas reviewed (changes requested):
Also, please resolve the merge conflicts.

PacoDw committed:
chore: removed groupID of the datasource
